File: 144_preorder_BST/recur_preorder.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> res;
    void preorder_trav(TreeNode* curr)
    {
        if (curr == NULL)
            return;

        res.push_back(curr->val);
        preorder_trav(curr->left);
        preorder_trav(curr->right);
    }
    vector<int> preorderTraversal(TreeNode* root) {
        if (root == NULL)
            return res;

        preorder_trav(root);
        return res;
    }
};

File: 144_preorder_BST/recur_preorder.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    trav_list = []

    def preorder_trav(self, curr):
        if curr == None:
            return
        
        self.trav_list.append(curr.val)
        self.preorder_trav(curr.left)
        self.preorder_trav(curr.right)
        
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        self.trav_list.clear()
        self.preorder_trav(root)
        return self.trav_list
        